Add support for eclipse.jdt.ls server

mkcms opened the pull request.

mkcms commented:
OK, now the only thing left to do is to use the `:workspaceFolders` thing properly. I thought it's not necessary to use it, but now I see that without it, the server cannot  function correctly (it gives the "classpath is incomplete" errors). I might have a solution.

mkcms commented:
> I'm not very happy with this. It seems the LSP server is doing something non-standard.

Yes, it is: https://github.com/eclipse/eclipse.jdt.ls/issues/376.
Basically, the server assumes (sigh...) that the client will apply the workspace edits manually when this code action is "invoked".

All the servers are currently doing it that way from what AFAIK. I believe that this change, or something similar that users can override should be included in eglot, because:

1. Even if we strictly follow the spec and don't override "java.apply.workspaceEdits" method for this specific server, the users don't have an easy way to hack it for themselves, if they want to have this non-standard thing. Currently I believe the only thing that they can do is advise `eglot-code-actions` on `jsonrpc-request` with some strange logic. We're complicating lives for users, who may not care about standards conformance. In the future, if we call `workspace/executeCommand` from more than one place, users will also have to advise these new commands.
2. We can use it in the future for other things. For example I have a code lens branch where commands are also called the same way. So by having one function we remove redundancy.  
3. Is it really that bad? There are two things here. First is the API `eglot-execute-command`. I don't really see any problem with having such a generic function. The second one is _overriding_ that function for the eclipse server. I agree, it may be ugly, but it makes the server usable. But even if we give up on the second thing, why complicate lives for our users and give up on the first? It's a very tiny change.

joaotavora commented on eglot.el:
```diff
@@ -1754,6 +1754,12 @@ If SKIP-SIGNATURE, don't try to send textDocument/signatureHelp."
                   "-data" workspace)))))
 
+(cl-defmethod eglot-execute-command
+  ((_server eglot-eclipse-jdt) command arguments)
```
Is `command` always a string? Because if it were a symbol (or maybe we could make it a symbol), we could use a `eql` defmethod specializer which would be quite a bit more elegant (no `cond` needed). I'll look into this, for now it can stay.

mkcms replied:
Yes, currently it's always a string. Or at least, the spec says it should be.

joaotavora commented:
> Is it really that bad?

It isn't.  I agree with all your arguments. It was kind of a knee-jerk reaction.  And I hadn't understood the problem.

Nice to know you have a codeLens branch...  You can, if you want, open a pull request for it even if it's not ready yet.

joaotavora commented:
> OK, now the only thing left to do is to use the :workspaceFolders thing properly.

I'll merge that "minimal" commit (if indeed it works) and also open an API for it (since advicing `project-roots` is kinda lame and it doesn't really sound like a project root, more like another non-standard hack).

mkcms commented:
I simplified the contact function a bit and added `workspaceFolders` detection in `initializationOptions`.

So far I've tested this with 5 Java projects popular on github and it seems to work. Currently, we assume that Java modules are located in project root directory and we add them to `workspaceFolders`. I also tried using `directory-files-recursively` to try and detect all possible modules (or subprojects), but this function signals `max-lisp-eval-depth` for me in nontrivial java projects.

What doesn't work is finding definitions of external dependencies (code completion, documentation, etc. works). I think to get that we would need to hack eglot some more, like @yyoncho said here: https://github.com/joaotavora/eglot/issues/63#issuecomment-412051303. (I don't wanna do that, except maybe in a separate package)

If you want to test this branch, you can simply do this:

1. Download and extract somewhere the server from http://download.eclipse.org/jdtls/snapshots/jdt-language-server-latest.tar.gz
2. Get a Java project, e.g. https://github.com/google/guava
3. Call `eglot` in some java file
4. When asked, point at the root directory where you extracted the server.

joaotavora commented:
> you can simply do this... You don't need to install eclipse or build it.

Thanks, that's much appreciated!

And thanks for all this work!

joaotavora commented:
Well I followed your instructions precisely and it kinda worked, so thanks.  I got prompted for the location of the LSP which was nice (and then it warned me my JAVA_HOME was not set, should it be? What do I do about this warning).

Then I started receiving a million unknown notifications in the events.

```
server-notification Sat Aug 11 16:21:41 2018:
(:jsonrpc "2.0" :method "language/status" :params
          (:type "Starting" :message "37% Starting Java Language Server - Importing Maven project(s)"))
```

On shutdown it did something non-standard again, I think and killed itself immediately after getting the `:shutdown` method.

Other than that, great job.  I'm going to rearrange your commits slightly by first cherry picking the infrastructure ones into master and then keeping the Java related ones in the PR.

Then I'll look into simplifying the contact function a bit if I can and run that by you.

mkcms commented:
> it warned me my JAVA_HOME was not set, should it be?

Well, it might work without it. If you install the JDK, it will try to deduce JAVA_HOME from path of `javac` executable. If it works without it though, maybe we can get rid of it altogether.

> Then I started receiving a million unknown notifications in the events.

Yes, that's expected. I also wonder if creating a temporary workspace is a good idea, because that way I think each project needs to be rebuilt and re-indexed every time the server is started. Maybe we should make a workspace in the project root, to speedup the startup.

joaotavora commented:
So I rebased this to master and simplified the eglot-execute-command thing.  I'll merge this tomorrow unless you want to work on this some more.

mkcms commented:
> I'll merge this tomorrow unless you want to work on this some more.

I wonder if we could somehow cache the path to the jar file? We could either modify CLASSPATH variable (I think users won't really expect that) or introduce another emacs variable or maybe something else entirely. It's your call.

mkcms commented:
> (is there some disadvantage to doing that?).

If you run an inferior process in Emacs, it will inherit this value. However this variable is AFAIK rarely used nowadays. I'll make a change to ask the user if they want to modify it.

joaotavora commented:
> However this variable is AFAIK rarely used nowadays. 

In that case, perhaps its better to do this:  let the path to the correct jar be in `eglot-server-programs`.  Then put all the current logic of the contact function into the `initialize-instance` method of the server class.  Read the docstring of `eglot-server-programs` to discover how this is done.  If you can't understand it, the docstring isn't well written enough :).  There's a chance that I messed something up and this isn't at all possible, but I think it is.

joaotavora commented:
@mkcms never mind my last comment.  The framework to do what I meant still needs some work

mkcms commented:
I wonder if we can have one (or a list) known location that we assume is the server extraction root. Then we would simply tell the user that we looked there and it wasn't found, and ask them for a path. We could also, in interactive call, ask the user if they want to store a symlink to the chosen path in the known location.

The path could be `~/.emacs.d/eclipse.jdt.ls` for example.

joaotavora commented:
@mkcms I've slept on this and decided the following: the place in eglot to customize the location of servers is `eglot-server-programs`.  If this LSP has such a complex invocation, then perhaps it shouldn't.  We shouldn't morph and twist Eglot into something to appease the whims of such a non-standard LSP, even if it is Java's, as that will only foster more non-conformities.  I will push for Eclipse JDT to include a startup script that does all this boring Java-related, most-definitely-non-eglot-related work, and in the meantime i think the CLASSPATH solution is fine.  Certainly there are other problems we can divert our attention to.

mkcms commented:
Alright, but I think we can still ask the user if they want to modify CLASSPATH to include the path to the jar there?

mkcms commented:
> OK, do that.

Done.

mkcms commented:
I changed contact function to point to one global workspace, as per #63.
I also rebased this onto master.

joaotavora commented:
I believe you need more people to test this. I admit I haven't tried much farther than following the recipe you gave before.  But it really didn't work very well.  Perhaps by Java environment is to blame.  Which JDK/JRE version do you recommend for that particular recipe?

mkcms commented:
> But it really didn't work very well.

What was wrong?

joaotavora commented:
> What was wrong?

[I told you](https://github.com/joaotavora/eglot/pull/65#issuecomment-412282182).  Apart from connecting, I couldn't really test any functionality in that "guava" project, it was perpetually stuck in that indexing phase.  Since I suspect it's already working slightly better for you, can you tell me which openjdk version you've installed (I remember that you mentioned openjdk, I think)?

mkcms commented:
> I told you. Apart from connecting, I couldn't really test any functionality in that "guava" project, it was perpetually stuck in that indexing phase.

I think it can be expected that the indexing will take a while since it's a pretty large project. How long did you wait? Did you not see any progress at all? I think that's a problem server-side, not eglot's itself.

Here's my output of `java -version`:

```
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-1~deb9u1-b11)
OpenJDK 64-Bit Server VM (build 25.171-b11, mixed mode)
```

By the way, the last change - creating one global workspace for all projects - will have to be reverted I think. A global workspace can get corrupted and the server will not start. It happened to me today. See here: https://github.com/eclipse/eclipse.jdt.ls/issues/622

joaotavora commented:
> I reverted the change that made the server use one global workspace.
> I reverted the change that made the server use one global workspace.

Hmm, isn't this a bit too drastic?  Can't we even reuse the workspace for the same project (one workspace per project, where the directory contains the project name)?

mkcms commented:
> Can't we even reuse the workspace for the same project (one workspace per project, where the directory contains the project name)?

Alright, I did that. I used `md5` for the directory name out of laziness, is it ok? I could also uniquify them by absolute path, like emacs auto save files put into one directory, but I didn't find a builtin function for it.

joaotavora commented:
I just rebase this to master. I intend to review it and merge it over the weekend. Sorry it's taking so long, but I hurt my hand and that makes programming a bit difficult :-(

joaotavora commented:
No sorry, not yet. Some additional personal stuff happened in the meantime, and I'm mostly away from my computer using my mobile to answer emails.

whatacold commented:
Hi @mkcms,

Thanks for your contribution.

> If you want to test this branch, you can simply do this:
> 
> 1. Download and extract somewhere the server from http://download.eclipse.org/jdtls/snapshots/jdt-language-server-latest.tar.gz. You don't need to install eclipse or build it.
> 2. Get a Java project, e.g. https://github.com/google/guava
> 3. Call `eglot` in some java file
> 4. When asked, point at the root directory where you extracted the server.

I followed you instructions with a maven project without luck, it complains:

    cl-no-applicable-method: No applicable method: project-roots, nil

Do I miss something?

P.S. I installed your fork with quelpa without any customization:

    (quelpa '(eglot :fetcher github :repo "mkcms/eglot" :branch "feature/eclipse.jdt.ls-server"))

mkcms commented:
@whatacold 
Thanks for reporting that.

> Do I miss something?

I think you're trying to start eglot in a buffer which is not in a project (that `project.el` can find). If you tried to start eglot in a buffer which is manage by git, I think it would work.

I pushed a commit which will handle the case when `project-current` returns nil. Can you check now?

mkcms commented:
By the way, you should know that you can only use this server in projects managed either by maven or gradle build systems. It will work in other projects also, but it will not provide many features (e.g. completion or diagnostics not related to syntax).

whatacold commented:
Hi @mkcms,

Thanks for your quick reply.

> I think you're trying to start eglot in a buffer which is not in a project (that project.el can find). If you tried to start eglot in a buffer which is manage by git, I think it would work.

I wasn't aware of project.el and wasn't familiar with it before, but it *is* a git controlled maven project.

After upgrading it with your new commit, it works now, it's great.

It has some warnings when `M-x eglot`, don't know whether it's expected or not:

    [eglot] Waiting in background for server `EGLOT (crawler4j/java-mode)'
    [eglot] (warning) Server sent unknown notification method `language/status' [2 times]
    [eglot] Connected! Server `EGLOT (crawler4j/java-mode)' now managing `java-mode' buffers in project `crawler4j'.
    [eglot] (warning) Server sent unknown notification method `language/status' [5 times]
    [eglot] (warning) Server sent unknown notification method `language/actionableNotification'

whatacold commented:
Another question about project root:)

Does the `crawler4j` in `[eglot] Waiting in background for server 'EGLOT (crawler4j/java-mode)'` mean the project root directory?
If it is, then I guess it's wrong, because it's not the directory where pom.xml resides.

See `:rootPath` property from the first few lines of buffer `*EGLOT (config/java-mode) events*`,
I'm pretty sure that it gets a wrong root path.

What do I need to do to tame it? I haven't figured it out yet after reading some bits of code.

whatacold commented:
Thanks a lot, it now detects the project root of maven correctly.

whatacold commented:
Hi @mkcms ,

Code action for inserting `import` statement for dos line ending file will insert an additional `^M` at the end.
For example, when I take the code action for `import java.util.Base64;`, the request and response in eglot events buffer are below,
where I replace the character `^M`(`\r`) with two characters `^M` for display purpose:
```
client-request (id:45) Mon Oct  1 19:31:49 2018:
(:jsonrpc "2.0" :id 45 :method "textDocument/codeAction" :params
          (:textDocument
           (:uri "file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App.java")
           :range
           (:start
            (:line 102 :character 10)
            :end
            (:line 102 :character 10))
           :context
           (:diagnostics
            [(:range
              (:start
               (:line 102 :character 8)
               :end
               (:line 102 :character 14))
              :severity 1 :code "570425394" :source "Java" :message "Base64 cannot be resolved")])))

server-reply (id:45) Mon Oct  1 19:31:50 2018:
(:jsonrpc "2.0" :id 45 :result
          [(:title "Import 'Base64' (java.util)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 2 :character 0)
                         :end
                         (:line 2 :character 0))
                        :newText "import java.util.Base64;")
                       (:range
                        (:start
                         (:line 2 :character 0)
                         :end
                         (:line 2 :character 0))
                        :newText "^M\n")]))])
           (:title "Import 'Base64' (org.apache.commons.codec.binary)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "^M\n^M\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "import org.apache.commons.codec.binary.Base64;")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "^M\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 5 :character 0))
                        :newText "")]))])
           (:title "Create constant 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "^M\n	")
                       (:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "private static final String Base64 = null;")]))])
           (:title "Create local variable 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "Object Base64;")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "^M\n		")]))])
           (:title "Change to 'Base32' (org.apache.commons.codec.binary)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "^M\n^M\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "import org.apache.commons.codec.binary.Base32;")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "^M\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 5 :character 0))
                        :newText "")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "Base32")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 14))
                        :newText "")]))])
           (:title "Change to 'BaseNode' (org.apache.james.mime4j.field.address)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 5 :character 30)
                         :end
                         (:line 5 :character 30))
                        :newText "^M\n")
                       (:range
                        (:start
                         (:line 5 :character 30)
                         :end
                         (:line 5 :character 30))
                        :newText "import org.apache.james.mime4j.field.address.BaseNode;")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "BaseNode")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 14))
                        :newText "")]))])
           (:title "Create field 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "^M\n	")
                       (:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "private Object Base64;")]))])
           (:title "Create parameter 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 29 :character 41)
                         :end
                         (:line 29 :character 41))
                        :newText ", ")
                       (:range
                        (:start
                         (:line 29 :character 41)
                         :end
                         (:line 29 :character 41))
                        :newText "Object Base64")]))])])
```

While the ones for linux line ending file are below:

```
client-request (id:97) Mon Oct  1 19:35:51 2018:
(:jsonrpc "2.0" :id 97 :method "textDocument/codeAction" :params
          (:textDocument
           (:uri "file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App.java")
           :range
           (:start
            (:line 102 :character 8)
            :end
            (:line 102 :character 14))
           :context
           (:diagnostics
            [(:range
              (:start
               (:line 102 :character 8)
               :end
               (:line 102 :character 14))
              :severity 1 :code "570425394" :source "Java" :message "Base64 cannot be resolved")])))

server-reply (id:97) Mon Oct  1 19:35:51 2018:
(:jsonrpc "2.0" :id 97 :result
          [(:title "Import 'Base64' (java.util)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 2 :character 0)
                         :end
                         (:line 2 :character 0))
                        :newText "import java.util.Base64;")
                       (:range
                        (:start
                         (:line 2 :character 0)
                         :end
                         (:line 2 :character 0))
                        :newText "\n")]))])
           (:title "Import 'Base64' (org.apache.commons.codec.binary)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "\n\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "import org.apache.commons.codec.binary.Base64;")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 5 :character 0))
                        :newText "")]))])
           (:title "Create constant 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "\n	")
                       (:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "private static final String Base64 = null;")]))])
           (:title "Create local variable 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "Object Base64;")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "\n		")]))])
           (:title "Change to 'Base32' (org.apache.commons.codec.binary)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "\n\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "import org.apache.commons.codec.binary.Base32;")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 3 :character 31))
                        :newText "\n")
                       (:range
                        (:start
                         (:line 3 :character 31)
                         :end
                         (:line 5 :character 0))
                        :newText "")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "Base32")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 14))
                        :newText "")]))])
           (:title "Change to 'BaseNode' (org.apache.james.mime4j.field.address)" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 5 :character 30)
                         :end
                         (:line 5 :character 30))
                        :newText "\n")
                       (:range
                        (:start
                         (:line 5 :character 30)
                         :end
                         (:line 5 :character 30))
                        :newText "import org.apache.james.mime4j.field.address.BaseNode;")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 8))
                        :newText "BaseNode")
                       (:range
                        (:start
                         (:line 102 :character 8)
                         :end
                         (:line 102 :character 14))
                        :newText "")]))])
           (:title "Create field 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "\n	")
                       (:range
                        (:start
                         (:line 27 :character 76)
                         :end
                         (:line 27 :character 76))
                        :newText "private Object Base64;")]))])
           (:title "Create parameter 'Base64'" :command "java.apply.workspaceEdit" :arguments
                   [(:changes
                     (:file:///path/to/crawler4j-try1/src/main/java/com/whatacold/crawler4j/App\.java
                      [(:range
                        (:start
                         (:line 29 :character 41)
                         :end
                         (:line 29 :character 41))
                        :newText ", ")
                       (:range
                        (:start
                         (:line 29 :character 41)
                         :end
                         (:line 29 :character 41))
                        :newText "Object Base64")]))])])
```

mkcms commented:
@whatacold 

Hmm, isn't that what is expected? Your file is using `\r\n` for newlines, so is the server. Or am I missing something? Can you provide the actual files?

whatacold commented:
Your words make me suspect if this problem came from my `.emacs.d`, so let me try it with `-Q`

1. bake a `/tmp/eglot-eclipse.el` to install your eglot fork with `quelpa`
2. start emacs with `emacs -Q -l /tmp/eglot-eclipse.el`
1. open a java file of any maven project
2. change it to dos line ending: `M-x set-buffer-file-coding-system RET dos RET`
3. input a new statement to trigger code actions of import, e.g. `Base64.getDecoder decoder;`
4. wait a few seconds, then click on `Base64` to pop up code action widget
5. select the one looking like `import "java.util"` to instruct eglot to insert `import java.util.Base64;`
6. the imported statement is like below, there is an additional `^M` at the end

```
import java.util.Base64;^M
import java.util.Set;
```

Here is my `/tmp/eglot-eclipse.el` :

```lisp
(require 'package)
;; you almost don't need this mirror
(setq package-archives
      '(("gnu" . "http://mirrors.163.com/elpa/gnu/")
        ("melpa" . "http://mirrors.163.com/elpa/melpa/")
        ("melpa-stable" . "http://mirrors.163.com/elpa/melpa-stable/")))

(setq package-user-dir "/tmp/elpa/")
(package-refresh-contents)
(package-initialize)

;; install the package you want to test and load it
(package-install 'quelpa)
(setq quelpa-dir "/tmp/quelpa")
(require 'quelpa)

(quelpa '(eglot :fetcher github :repo "mkcms/eglot" :branch "feature/eclipse.jdt.ls-server"))

(defun whatacold/eglot-eclipse-jdt-ls-jar-set (directory)
  "Set CLASSPATH env for eglot and eclipse jdt ls, if DIRECTORY is the root directory."
  (let* ((regex "org\\.eclipse\\.equinox\\.launcher_.*\\.jar$")
         (jar (directory-files (concat directory "plugins/") t regex)))
    (when (listp jar)
      (setenv "CLASSPATH" (concat (getenv "CLASSPATH") ":" (car jar))))))

(whatacold/eglot-eclipse-jdt-ls-jar-set (concat user-emacs-directory "bin/eclipse.jdt.ls/"))
```

mkcms commented:
@whatacold 
I can't reproduce the problem with your recipe. Every time I do it, I don't see any extra characters. 
By the way, after step 4 you should save the buffer to actually make sure the server knows about the changes, then revert that buffer, maybe that will help.

Can you send the actual file you are editing right before the step 6?

whatacold commented:
@mkcms 

Too bad to hear that it can't be reproduced, and I can't always reproduce it too :(

I'll create a github repo if I find out how to reproduce it steadily.

whatacold commented:
@mkcms 

Though don't know how to reproduce yet, but I think this issue stems from the ls server,
because the normal `textDocument/codeAction` response is `:newText "\n\n"`,
while the abnormal one is `:newText "^M\n^M\n"`.

Even for a dos file, the content in `textDocument/didSave` notification sent to server contains `\n` only,
as I observed in the `M-x eglot-event-buffer` . So I think ls don't know what the line ending style is.

I tried to find the answer from eclips.jdt.ls, but haven't figured out yet.

mkcms commented:
I intend to add tests for this PR later this week.

joaotavora commented:
@mkcms there is a failure on travis, on the master branch.  Even though the test is `rls`'s it seems to be suffering interference from a previous `eclipse-jdt` run.  

Can you investigate?

PS: In the future, it might be better to first merge the PR into a copy of master, let the tests run there, and then really commit to the merge to master.




```
Test rls-rename backtrace:
  signal(error ("[eglot] -32603: Internal error."))
  error("[eglot] %s" "-32603: Internal error.")
  eglot--error("-32603: Internal error.")
  #f(compiled-function (&rest rest) #<bytecode 0xfea8b9>)(:code -32603
  apply(#f(compiled-function (&rest rest) #<bytecode 0xfea8b9>) (:code
  #f(compiled-function (jsonrpc-lambda-elem7) #<bytecode 0xe9f811>)((:
  jsonrpc-connection-receive(#<eglot-eclipse-jdt eglot-eclipse-jdt> (:
  jsonrpc--process-filter(#<process EGLOT (project/java-mode)> "96)\\n
  accept-process-output(nil 30)
  #f(compiled-function (arg1 arg2 arg3 &rest rest) "Make a request to 
  apply(#f(compiled-function (arg1 arg2 arg3 &rest rest) "Make a reque
  jsonrpc-request(#<eglot-rls eglot-rls> :textDocument/rename (:textDo
  eglot-rename("bla")
  #f(compiled-function () #<bytecode 0xe7f0b5>)()
  eglot--call-with-dirs-and-files((("rename-project" ("main.rs" . "fn 
  #f(compiled-function () #<bytecode 0xe7f0c1>)()
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name rls-rename :documentation "Test renam
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval((ert-run-tests-batch-and-exit 't))
  command-line-1(("-L" "." "--eval" "(package-initialize)" "--eval" "(
  command-line()
  normal-top-level()
```

mkcms commented:
I tried fixing the tests on CI and so far I can't do it. On my machine they always pass now, but on travis they fail.

The problem I think is that `eglot--with-dirs-and-files` hooks into `eglot-connect-hook` to detect server connection in `BODY`, but that hook is called after `BODY` somehow.

joaotavora commented:
Ok, thanks.  I'll see what I can do.  In the meantime, please temporarily disable the eclipse test in master.
